File: apps/api/src/models/goal.ts
import { AggregateRoot } from "@abejarano/ts-mongodb-criteria";
import { createMongoId } from "./shared/mongo_id";

export type GoalPrimitives = {
  id?: string;
  goalId: string;
  userId: string;
  name: string;
  targetAmount: number;
  currency: string;
  startDate: Date;
  targetDate?: Date;
  monthlyContribution?: number;
  linkedAccountId?: string;
  isActive: boolean;
  createdAt: Date;
  updatedAt?: Date;
};

export type GoalCreateProps = {
  userId: string;
  name: string;
  targetAmount: number;
  currency?: string;
  startDate: Date;
  targetDate?: Date;
  monthlyContribution?: number;
  linkedAccountId?: string;
  isActive?: boolean;
};

export class Goal extends AggregateRoot {
  private readonly props: GoalPrimitives;

  private constructor(props: GoalPrimitives) {
    super();
    this.props = props;
  }

  static create(props: GoalCreateProps): Goal {
    const now = new Date();
    return new Goal({
      goalId: createMongoId(),
      userId: props.userId,
      name: props.name,
      targetAmount: props.targetAmount,
      currency: props.currency ?? "BRL",
      startDate: props.startDate,
      targetDate: props.targetDate,
      monthlyContribution: props.monthlyContribution,
      linkedAccountId: props.linkedAccountId,
      isActive: props.isActive ?? true,
      createdAt: now,
      updatedAt: now,
    });
  }

  static fromPrimitives(primitives: GoalPrimitives): Goal {
    return new Goal(primitives);
  }

  toPrimitives(): GoalPrimitives {
    return this.props;
  }

  getId(): string {
    return this.props.id ?? this.props.goalId;
  }

  getGoalId(): string {
    return this.props.goalId;
  }
}

File: apps/flutter/lib/features/goals/domain/entities/goal.dart
class Goal {
  final String id;
  final String name;
  final double targetAmount;
  final String currency;
  final DateTime startDate;
  final DateTime? targetDate;
  final double? monthlyContribution;
  final String? linkedAccountId;
  final bool isActive;

  const Goal({
    required this.id,
    required this.name,
    required this.targetAmount,
    required this.currency,
    required this.startDate,
    required this.targetDate,
    required this.monthlyContribution,
    required this.linkedAccountId,
    required this.isActive,
  });

  factory Goal.fromJson(Map<String, dynamic> json) {
    return Goal(
      id: (json["goalId"] ?? json["id"]) as String,
      name: json["name"] as String,
      targetAmount: (json["targetAmount"] as num?)?.toDouble() ?? 0,
      currency: json["currency"] as String? ?? "BRL",
      startDate: DateTime.parse(json["startDate"] as String),
      targetDate: json["targetDate"] != null
          ? DateTime.parse(json["targetDate"] as String)
          : null,
      monthlyContribution:
          (json["monthlyContribution"] as num?)?.toDouble(),
      linkedAccountId: json["linkedAccountId"] as String?,
      isActive: json["isActive"] as bool? ?? true,
    );
  }
}
